File: Java-Core/src/main/java/presto/value/ContextualExpression.java
package presto.value;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.type.IType;
import presto.utils.CodeWriter;

/* a value which can only be determined when expression is interpreted in specific context */
public class ContextualExpression extends BaseValue implements IExpression {

	Context calling;
	IExpression expression;
	
	public ContextualExpression(Context calling, IExpression expression) {
		super(null); // TODO check that this is not a problem
		this.calling = calling;
		this.expression = expression;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		expression.toDialect(writer);
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		return expression.check(this.calling);
	}

	@Override
	public IValue interpret(Context context) throws PrestoError {
		return expression.interpret(this.calling);
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/value/ContextualExpression.js
var Value = require("./Value").Value;

function ContextualExpression(calling, expression) {
	Value.call(this, null); // TODO check that this is not a problem
	this.calling = calling;
	this.expression = expression;
	return this;
}

ContextualExpression.prototype = Object.create(Value.prototype);
ContextualExpression.prototype.constructor = ContextualExpression;

ContextualExpression.prototype.toDialect = function(dialect) {
	return this.expression.toDialect(dialect);
};

ContextualExpression.prototype.check = function(context) {
	return this.expression.check(this.calling);
};

ContextualExpression.prototype.interpret = function(context) {
	return this.expression.interpret(this.calling);
};

exports.ContextualExpression = ContextualExpression;